### PR TITLE
Support custom jumbo type

### LINKED ISSUE #390: Deserialization should not rely on classes with constructors having large parameter counts
## Hello!

- Vote on this issue by adding a 👍 reaction
- If you want to implement this feature, comment to let us know (we'll work with you on design, scheduling, etc.)

## Issue details

Deserialization should not rely on classes with constructors having large parameter counts.

Code generation for Java SDK classes that need to be deserialized from protobuf currently relies on constructors with many parameters (one per desired property) internally. This breaks down when it hits hard limits on Java parameter counts. 

Earlier when we worked around the problem of not being able to compile providers with large parameter counts, we introduced a special pathway in the SDK code generator for the special case of "jumbo" constructors. This does not completely solve the problem however. While you can construct these Jumbo objects, tryConvertObjectInner errors when facing them.

Desired complete fix would:

- [x] fix the deserializer in the Java SDK to not require one parameter per property in any method or constructor of the generated classes, so as to not hit the Java limit
- [ ] fix the codegen to align with the deserializer expectations
- [ ] remove branching on parameter count from codegen code
- [ ] have no breaking changes for the users (it seems the problem can be solved internally)
- [ ] fix or avoid "`Converter::tryConvertObjectInner` errors when calling jumbo constructors"  found earlier

### PR DESCRIPTION
**Pull request by pawelprazak**
# Description

Fixes #390 

## Checklist

- [ ] I have added tests that prove my fix is effective or that my feature works

- [x] I have updated the [CHANGELOG-PENDING](https://github.com/pulumi/pulumi/blob/master/CHANGELOG_PENDING.md) file with my change

- [ ] Yes, there are changes in this PR that warrants bumping the Pulumi Service API version

### CODE REVIEW

**Comment by pawelprazak** [comment]
Rebased.

WDYT, should I add more tests, or it this good enough?

File: sdk/java/pulumi/src/test/java/com/pulumi/serialization/internal/JumboCustomType.java [hunk 1]
```diff
@@ -0,0 +1,3111 @@
...
+    	      this.bar93 = defaults.bar93;
+    	      this.bar94 = defaults.bar94;
+    	      this.bar95 = defaults.bar95;
+    	      this.bar96 = defaults.bar96;
+    	      this.bar97 = defaults.bar97;
+    	      this.bar98 = defaults.bar98;
+    	      this.bar99 = defaults.bar99;
+        }
+
+        @CustomType.Setter
```
Comment (t0yv0) [inline sdk/java/pulumi/src/test/java/com/pulumi/serialization/internal/JumboCustomType.java:1568]: Why does it need both @CustomType.Setter and @CustomType.Parameter - curious?

File: CHANGELOG_PENDING.md [hunk 2]
```diff
@@ -1,3 +1,5 @@
 ### Improvements
 
+- [codegen] #390: support resources with over 255 parameters
```
Comment (t0yv0) [inline CHANGELOG_PENDING.md:3]: Trying to remember if this was unique to resources - I think the issue applies to all generated classes not just resources.

**Comment by t0yv0** [comment]
I would expect to see changes to pkg/codegen/java/gen.go - are those pending or what's the plan here?

- removal of `genJumboOutputType`
- removal of `IsJumbo`

**Comment by t0yv0** [comment]
Java code seems reasonable. We can tune it for performance later if the ser/de ever becomes a problem - I'm guessing reflective invocations are not the fastest.

File: CHANGELOG_PENDING.md (hunk 2 above)
Comment (pawelprazak) [reply CHANGELOG_PENDING.md:3]: The mechanism being changed works only for custom resources specifically. Anything annotated with `@CustomType`.

**Comment by pawelprazak** [comment]
> I would expect to see changes to pkg/codegen/java/gen.go - are those pending or what's the plan here?
> 
> * removal of `genJumboOutputType`
> * removal of `IsJumbo`

I've assumed we don't want to break current providers and I've planned to remove the constructor-based code, after we release both the new codegen and regenerate all of the (affected) providers.

As a result, I plan to do as follows, in 4 stages:
- SDK change (this PR), release the SDK
- update codegen to use the new mechanism
- regenerate and release providers with the new codegen
- remove the old mechanism from codegen and from SDK

**Comment by pawelprazak** [comment]
updated after rebase

File: CHANGELOG_PENDING.md (hunk 2 above)
Comment (t0yv0) [reply CHANGELOG_PENDING.md:3]:
```
/**
 * Annotation used by a Pulumi Cloud Provider Package to mark complex types used for a Resource
 * output property.
 * <p>
 * A complex type must have a single constructor
 * marked with the @see {@link CustomType.Constructor} annotation.
 */
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface CustomType {
```

Ok. Judging by the comment, the types annotated with this are not modeling Pulumi resources. They are the types used to describe outputs of resources. It sounds like the fix is doing the right thing but the CHANGELOG entry is misleading, that's no problem, I can update later on as it comes time to release.

**Review by t0yv0 — APPROVED** [review]

File: sdk/java/pulumi/src/test/java/com/pulumi/serialization/internal/JumboCustomType.java (hunk 1 above)
Comment (t0yv0) [reply sdk/java/pulumi/src/test/java/com/pulumi/serialization/internal/JumboCustomType.java:1568]: Please be mindful of gigantic generated code sizes that remain a large problem for the speed of Pulumi experience. If something can be solved with fewer API elements we should probably consider that as it puts less pressure on the generated code. I don't have the numbers so perhaps this is negligible, but you could theoretically take into account that setters will only have 1 parameter, and record the name of said parameter in `@CustomType.Setter`. You could also skip recording the name if it happens to match a reflectable name. 

That said I'm good with merging this for now, just consider generated code size when making further changes in the future.

**Comment by t0yv0** [comment]
LGTM but could we please track all the remaining work? In #390 feel free to edit the checklist as needed. Thank you!

File: sdk/java/pulumi/src/test/java/com/pulumi/serialization/internal/JumboCustomType.java (hunk 1 above)
Comment (pawelprazak) [reply sdk/java/pulumi/src/test/java/com/pulumi/serialization/internal/JumboCustomType.java:1568]: > You could also skip recording the name if it happens to match a reflectable name.

Unfortunately this is not possible, we change the name of the setter in codegen in some cases.

Yes, you are right, I could reuse the same annotation. I'll make a followup PR with this.

Comment (pawelprazak) [reply sdk/java/pulumi/src/test/java/com/pulumi/serialization/internal/JumboCustomType.java:1568]: the follow up: https://github.com/pulumi/pulumi-java/pull/730

**Pull request merged**